### PR TITLE
〈New〉import command.

### PR DESCRIPTION
**Pull request by bucaran**
### Breakdown

+ Rewrites `oh-my-fish.fish` initial configuration and adds in-depth comments. The amount of code is reduced to ~__10__ lines of code.

+ Adds a __new__ [`import`](https://github.com/bucaran/oh-my-fish/blob/import/functions/import.fish) command and introduces the idea of _libraries_. A library is category to  conceptually organize cluster of `.fish` files such as plugins, themes and completions. A plugin intended to be used by other plugins and not by the user of the shell can be thought of as a _library_. 

+ Simplifies `_prepend_tree` reducing amount of code to half.

+  `import` calls [`_prepend_tree`](https://github.com/bucaran/oh-my-fish/blob/import/functions/_prepend_tree.fish) for each plugin in the global `$fish_plugins` prepending subdirectories containing `.fish` files to the `$fish_function_path`. 

Now it's possible to create complex directory trees with `fish` files organized according to the plugin's own API / structure. Paths containing `.fish` files will be added to the path automatically. This should encourage plugin authors to create a consistent project hierarchy without having to worry with load / configuration files modifying the `$fish_function_path` directly. `.load` files can still be used for more pertinent tasks such as downloading assets, declaring global variables, etc. 

> __Note__: `_prepend_tree` runs before `.load` files are sourced, so it is still possible to override any undesired behavior in the `<plugin>.load` file if necessary.

### Notes

`import` replaces `oh_my_fish.fish`, `_fish_add_plugin`, `_fish_add_completion`, `_fish_source_plugin_load_file` functions inside `functions/oh-my-fish.fish/`, adds documentation and calls `_prepend_tree`, but it is still less code than the current configuration. This is possible using Fish `{}` brace expansion:

```fish
    set -l root $fish_path $fish_custom
    # ...
    set -l paths $root/$library $root/$library/completions
    set -l plugins     $paths[1..2]" fish_function_path"
    set -l completions $paths[3..4]" fish_complete_path"
    
    for path in $plugins $completions
      eval "_prepend_path "$path
    end
```

#### Performance
I tested adding _al_ l the existing modules with a macbook pro and while I did notice a small lag when starting the shell the first time, the bottleneck seems to be simply adding so many plugins (and the fact Fish [can not](https://github.com/fish-shell/fish-shell/issues/563) run Fish code in the background) and not `_prepend_tree` running for each plugin, conclusion I reached after commenting it and retesting without being able to notice any significant performance gains. See #96 for more info.

#### `import <path/library>[<path/library>..]`
Imports libraries, plugins, themes, completions. Prepends existing user `custom/<library>` directories to the path to  allow users to override specific functions. Use `import` to load plugins, themes, handle plugin inter-dependency, etc. 

> __Note__: Adds __only__ existing directories to the path.

### Examples
+ `import plugins/dpaste themes/bobthefish`
+ `import plugins/{cask,brew,django}`

### Extras

+ Also included in this commit is the [`msg`](https://github.com/bucaran/oh-my-fish/blob/import/plugins/msg/msg.fish) plugin. See #285.
msg

### Related & Other Issues

[\#286](https://github.com/bpinto/oh-my-fish/pull/286) [\#284](https://github.com/bpinto/oh-my-fish/pull/284) #285 #96

### CODE REVIEW

**Comment by bpinto** [comment]
Let's start the review... :)

**Comment by bucaran** [comment]
Take your time and let me know anything.

File: functions/_prepend_tree.fish [hunk 1]
```diff
@@ -1,72 +1,42 @@
...
-#             -a -and
-#                 Any matches must meet all listed criteria.
-#
-#                     glob1 -a glob2  →  glob1 AND glob2
-#
-#      -d <DESTINATION PATH>
-#          Should appear at the end if used. Specifies the name of the
-#          global path variable to prepend the matched directories to.
-#          If not used, the $fish_function_path is assumed by default.
+#      [glob1 ] -a -and [glob2]
```
Comment (bpinto) [inline functions/_prepend_tree.fish]: `[glob1] [-o -or] [glob2] ...`
`[glob1 ] -a -and [glob2]`

could share the same pattern.

`[<glob> [<operator> <glob>..]]` could possibly too.

Comment (bucaran) [inline functions/_prepend_tree.fish]: I agree, however the reason for `[-o -or]`, but only `-a -and` is that the `or` operator is optional (used by default if no operator is used), whereas `and` must be specified if you are planning to use it.

File: functions/import.fish [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    # Prepend plugins and completion paths to fish_function_path and
+    # fish_complete_path respectively. The root paths are determined
+    # via the global $fish_path and $fish_custom.
+    for path in $plugins $completions
+      eval "_prepend_path "$path
+    end
+
+    # Traverse the library tree and prepend directories with fish files.
+    _prepend_tree $fish_path/$library
```
Comment (bpinto) [inline functions/import.fish]: Isn't this doing the same thing as 
```
+    set -l plugins     $paths[1..2]" fish_function_path"
+    for path in $plugins $completions
+      eval "_prepend_path "$path
+    end
```

File: functions/import.fish [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    # Source each library (plugin, theme,...) configuration load file.
+    for path in $root/$library/(basename $library).load
+      [ -e $path ]
+        and eval ". "$path
```
Comment (bpinto) [inline functions/import.fish]: Not very fond of `eval`, I'd write this as:

```
for path in $root/$library/(basename $library).load
  if test -e $path
    . $path
  end
end
```

Comment (bucaran) [inline functions/import.fish]: Agreed and good catch. Will fix it right away.

File: functions/import.fish (hunk 2 above)
Comment (bucaran) [inline functions/import.fish]: What do you mean by _this_? Why am I seeing a part of the code detached from the function?

File: oh-my-fish.fish [hunk 4]
```diff
@@ -1,74 +1,32 @@
...
+# Initial Configuration
+##
 
 # Set fish_custom to the path where your custom config files
-# and plugins exists, or else we will use the default custom.
-if not set -q fish_custom
-  set -g fish_custom  $fish_path/custom
-end
+# and plugins exist, or use the default custom instead.
+not set -q fish_custom
```
Comment (bpinto) [inline oh-my-fish.fish]: The `if` version is more readable.

File: oh-my-fish.fish [hunk 5]
```diff
@@ -1,74 +1,32 @@
...
-end
-
-# Add all defined plugins
-for plugin in $fish_plugins
-  _fish_add_plugin $plugin
-  _fish_add_completion $plugin
-  _fish_source_plugin_load_file $plugin
-end
+# Add functions defined in oh-my-fish/functions to path.
+not contains $fish_path/functions/ $fish_function_path
```
Comment (bpinto) [inline oh-my-fish.fish]: The `if` version is more readable.

File: functions/import.fish (hunk 3 above)
Comment (bucaran) [inline functions/import.fish]: Oh wait, you do realize that I need to use `eval` because I am doing brace expansion. That's why we are able to simplify a lot.

File: oh-my-fish.fish (hunk 4 above)
Comment (bucaran) [inline oh-my-fish.fish]: I don't have a strong opinion on this. I prefer not using `if` when one liners are possible, and use `if` for > 1 liners. I can revert it if you really want to.

File: plugins/rbenv/rbenv.load [hunk 6]
```diff
@@ -1,7 +1,11 @@
-set -l rbenv_dir "$RBENV_ROOT"
+set -l rbenv_dir "$ec"
```
Comment (bpinto) [inline plugins/rbenv/rbenv.load]: Another pull request?

File: plugins/msg/msg.fish [hunk 7]
```diff
@@ -0,0 +1,472 @@
+#    __ _  ___ ___ _
```
Comment (bpinto) [inline plugins/msg/msg.fish]: Separate msg plugin in another pr?

File: functions/_prepend_tree.fish (hunk 1 above)
Comment (bpinto) [inline functions/_prepend_tree.fish]: :+1:

File: plugins/rbenv/rbenv.load (hunk 6 above)
Comment (bucaran) [inline plugins/rbenv/rbenv.load]: :grinning: It was crashing my terminal when I tried loading the plugin so I figured it is loosely related, considering this PR modifies a lot of the foundational elements in the framework.

File: plugins/msg/msg.fish (hunk 7 above)
Comment (bucaran) [inline plugins/msg/msg.fish]: Let me submit another PR for import first.

File: functions/import.fish (hunk 2 above)
Comment (bpinto) [inline functions/import.fish]: _this_ is the line of code that comment was added.

Thorough example:

Considering my plugin is called: _my_plugin_
```
set -l root $fish_path $fish_custom
set -l paths $root/$library $root/$library/completions
set -l plugins     $paths[1..2]" fish_function_path"

for path in $plugins $completions
  echo "_prepend_path "$path
end
```

Output:

`_prepend_path /Users/bpinto/.oh-my-fish/my_plugin fish_function_path`
`_prepend_path /Users/bpinto/Dropbox/Documents/Bruno/dotfiles/oh-my-fish/my_plugin fish_function_path`


The first `_prepend_path` would be the included in `_prepend_tree $fish_path/$library`. So I don't think we need that `_prepend_path` logic for plugins, only for completions. Or did I miss something? :confused:

File: oh-my-fish.fish (hunk 4 above)
Comment (bpinto) [inline oh-my-fish.fish]: I'd feel more comfortable. I think people are more used to write an `if` clause in here instead of using boolean logic.

File: plugins/rbenv/rbenv.load (hunk 6 above)
Comment (bpinto) [inline plugins/rbenv/rbenv.load]: I think this will break for other people, I don't have $ec defined on my computer for instance. So, I guess this would not work for me.

File: oh-my-fish.fish (hunk 4 above)
Comment (bucaran) [inline oh-my-fish.fish]: Cool, let's use the `if` version. I hope you let me get away with the one liners in my own plugins though :smiley:

Comment (bpinto) [inline oh-my-fish.fish]: I hope I can convince everything is owned by the community. :P

File: plugins/rbenv/rbenv.load (hunk 6 above)
Comment (bucaran) [inline plugins/rbenv/rbenv.load]: I don't use `rbenv` so I don't mind ignoring this, but I don't understand why it would break for other people. It will only _not_ break for people without `rbenv`, like me.

File: oh-my-fish.fish (hunk 4 above)
Comment (bucaran) [inline oh-my-fish.fish]: Of course you know I mean the plugins I wrote, like `msg`. :stuck_out_tongue_closed_eyes:

File: functions/import.fish (hunk 3 above)
Comment (bpinto) [inline functions/import.fish]: I don't see why this wouldn't work in here. Can you explain it to me?

File: oh-my-fish.fish (hunk 4 above)
Comment (bpinto) [inline oh-my-fish.fish]: :+1:

File: functions/import.fish (hunk 3 above)
Comment (bucaran) [inline functions/import.fish]: My bad, we don't need `eval` for the last loop. We _do_ need it for the one with brace expansion.

`eval "_prepend_path "$path`

Comment (bpinto) [inline functions/import.fish]: :+1:

File: functions/import.fish (hunk 2 above)
Comment (bpinto) [inline functions/import.fish]: Having said that, what about `_prepend_tree $fish_path/$library` for `$fish_custom`?

Comment (bucaran) [inline functions/import.fish]: Excellent catch.

Useless and inconsequential trivia: actually _not_ in the original `_prepend_tree` version, but later I changed the main loop in _prepend_tree` to

```fish
  for dir in $path $path/**/
  ...
```

effectively including the same directory and making you right about this.

File: plugins/rbenv/rbenv.load (hunk 6 above)
Comment (bpinto) [inline plugins/rbenv/rbenv.load]: Why would this be executed if you don't have rbenv? I assume you don't have it set on your $fish_plugins.

That line is using my exported $RBENV_ROOT environment variable to define the rbenv dir. If we remove that, then it won't work for my customized install dir.

Comment (bucaran) [inline plugins/rbenv/rbenv.load]: It will crash my terminal if I try to load the plugin when not having `rbenv`.

Comment (bpinto) [inline plugins/rbenv/rbenv.load]: I don't know what happen to the other plugins if you don't have the dependencies installed.
I think this is fine.

**Comment by bpinto** [comment]
I'll do some debugging, it's noticeable slower to start my terminal with this branch.

File: plugins/rbenv/rbenv.load (hunk 6 above)
Comment (bucaran) [inline plugins/rbenv/rbenv.load]: This was the only one giving trouble in my machine, but you are right, we could improve this later. Will remove in the next `import` PR.

File: functions/import.fish (hunk 2 above)
Comment (bucaran) [inline functions/import.fish]: __However__, we still want to include `$plugins` in the first loop in order to add user custom directories that can override plugin functionality. The fix, is changing:

`set -l plugins     $paths[1..2]" fish_function_path"` __to__ `set -l plugins     $paths[2]" fish_function_path"`.

The first index refers to `$root/$library` which is the actual plugin directory already added by `_prepend_tree`.

**Comment by bucaran** [comment]
Updated PR.

**Comment by bpinto** [comment]
Just tried, still not as fast. I believe the `eval` is causing it, let me try one thing.

**Comment by bucaran** [comment]
I got it.

File: plugins/rbenv/rbenv.load [hunk 8]
```diff
@@ -1,3 +1,8 @@
...
```
Comment (bpinto) [inline plugins/rbenv/rbenv.load]: There is still a bug on this code, don't worry about rbenv for now.

```
.oh-my-fish:prepend_path λ set rbenv_dir abc
.oh-my-fish:prepend_path λ echo $rbenv_dir
abc
.oh-my-fish:prepend_path λ set rbenv_dir "$I_DONT_EXIST"
.oh-my-fish:prepend_path λ echo $rbenv_dir

```

**Comment by bucaran** [comment]
Just pushed new changes. It seems to be loading just as fast as your `master`.

File: functions/_prepend_tree.fish [hunk 9]
```diff
@@ -74,68 +44,48 @@
...
+  # Travese $path and prepend only directories with matches.
+  for dir in "$path" "$path"/**
+    # Use head to retrieve at least the first match.
+    if [ -z (find $dir $glob -maxdepth 1 | head -1) ]
+      continue
+    end
+    if contains -- $argv[1] -p --preview
+      printf "%s\n" $dir
+    else
+      _prepend_path $dir fish_function_path
```
Comment (bpinto) [inline functions/_prepend_tree.fish]: https://github.com/bpinto/oh-my-fish/pull/290 alter this.

File: functions/import.fish [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  for library in $argv
+    # Prepend plugins, themes and completions, as well as user custom paths
+    # for plugins and themes to $fish_function_path and $fish_complete_path.
+    # The root paths are determined via $fish_path and $fish_custom globals.
+    for path in $fish_custom
+      _prepend_path $path fish_function_path
+    end
+    
+    for path in $fish_path/$library/completions
+      _prepend_path $path fish_function_path
```
Comment (bpinto) [inline functions/import.fish]: This should be `fish_complete_path`

File: functions/import.fish [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+#      functions/_prepend_tree.fish
+#
+# v.0.1.1
+#/
+function import -d "Load libraries, plugins, themes, etc."
+  for library in $argv
+    # Prepend plugins, themes and completions, as well as user custom paths
+    # for plugins and themes to $fish_function_path and $fish_complete_path.
+    # The root paths are determined via $fish_path and $fish_custom globals.
+    for path in $fish_custom
```
Comment (bpinto) [inline functions/import.fish]: `$fish_custom/$library`

File: functions/_prepend_tree.fish (hunk 9 above)
Comment (bucaran) [inline functions/_prepend_tree.fish]: So, how do you proceed here? Do you merge your commit and then the import branch? Or do you want me to alter anything?

File: functions/import.fish (hunk 11 above)
Comment (bpinto) [inline functions/import.fish]: This does not allow custom plugins to have subfolders, there is no reason for that.

File: functions/import.fish (hunk 10 above)
Comment (bpinto) [inline functions/import.fish]: With the new `_prepend_path` we would be able to call `_prepend_path $fish_path/$library/completions -d fish_function_path` instead.

File: functions/_prepend_tree.fish (hunk 9 above)
Comment (bpinto) [inline functions/_prepend_tree.fish]: I think it would be better as we can simplify the calls above. Otherwise I'd merge that #290 later.

File: functions/import.fish (hunk 11 above)
Comment (bucaran) [inline functions/import.fish]: Subfolders are handled by `_prepend_tree $fish_path/$library`

Comment (bpinto) [inline functions/import.fish]: But if they are custom, they are not under `$fish_path`, but `$fish_custom`.
I think we need two `_prepend_tree` calls, one for `$fish_path` and another for `$fish_custom`.

File: functions/_prepend_tree.fish (hunk 9 above)
Comment (bpinto) [inline functions/_prepend_tree.fish]: Merged #290.

File: functions/import.fish (hunk 11 above)
Comment (bpinto) [inline functions/import.fish]: I think we are talking about different things.

```
# I have a custom boot2docker plugin
.oh-my-fish:master λ echo $fish_plugins
brew boxen boot2docker bundler cliniko dpaste extract ndenv osx rails z rbenv

# My custom folder is not under ~/.oh-my-fish directory.
.oh-my-fish:master λ echo $fish_custom
/Users/bpinto/Dropbox/Documents/Bruno/dotfiles/oh-my-fish

# The custom plugins has a nested folder
.oh-my-fish:master λ ls $fish_custom/plugins/boot2docker/
boot2docker.load clean.fish       nested_folder
```

Comment (bucaran) [inline functions/import.fish]: Ah, I see!

File: functions/import.fish (hunk 10 above)
Comment (bucaran) [inline functions/import.fish]: You mean of course replacing both loops for:
`_prepend_path $fish_path/$library -d fish_function_path`
 `_prepend_path $fish_path/$library/completions -d fish_complete_path`

Comment (bpinto) [inline functions/import.fish]: :+1:

Comment (bucaran) [inline functions/import.fish]: Because of `_prepend_tree` we don't need `_prepend_path $fish_path/$library -d fish_function_path`.

**Comment by bucaran** [comment]
This merges your changes, and updates calls to the new `_prepend_path`.

**Comment by bucaran** [comment]
I added another small improvement during the initial setup.

```fish
# Prepend all user custom paths to the fish path and source load files. 
for custom_file in $fish_custom/**
  _prepend_path $custom_file -d fish_function_path
  switch $custom_file
    case \*.load
      . $custom_file
  end
end
```

In the same loop we are using to source the .load files we can also `_prepend_path` only those that are directories.

**Comment by bpinto** [comment]
I'll to another full review, but if you could squash and rebase, that would be the next step.

File: functions/_prepend_tree.fish [hunk 12]
```diff
@@ -74,68 +44,48 @@
...
 
-  [ -z "$glob" ]
-    and set glob -name \*.fish
-
-  for directory in $source/**/
-    if not [ -z (find $directory $glob -maxdepth $depth | head -1) ]
-      eval $preview # skip when on preview mode
-        and printf "%s " $directory
-        or _prepend_path $directory $destination
+  # Travese $path and prepend only directories with matches.
```
Comment (bpinto) [inline functions/_prepend_tree.fish]: Typo traverse.

File: oh-my-fish.fish [hunk 13]
```diff
@@ -1,74 +1,34 @@
...
+# customized via the $fish_path/custom directory. To customize a theme,
+# create a directory under $fish_path/custom/themes with the same name
+# as the theme. Use the same approach for plugins, etc.
+import plugins/$fish_plugins themes/$fish_theme
 
-# Source all files inside custom folder
-for config_file in $fish_custom/*.load
-  . $config_file
+# Prepend all user custom paths to the fish path and source load files. 
+for custom_file in $fish_custom/**
```
Comment (bpinto) [inline oh-my-fish.fish]: I don't understand why custom folder is handled differently. I could add a dependency from a custom plugin to another custom plugin.

File: functions/import.fish [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+  for library in $argv
+
+    # Prepend plugins, themes and completions. Also traverse the library
+    # tree prepending directories with fish code. This should encourage
+    # to create project trees with fish code organized in directories.
+    _prepend_tree $fish_path/$library
+    _prepend_path $fish_path/$library/completions -d fish_complete_path
+
+    # Source each library (plugin, theme,...) configuration load file.
+    for path in $root/$library/(basename $library).load
```
Comment (bpinto) [inline functions/import.fish]: Root is undefined.

File: oh-my-fish.fish (hunk 13 above)
Comment (bpinto) [inline oh-my-fish.fish]: By not using the same handler (import function) we are also not filtering the custom plugins using the fish_plugins env var

File: functions/import.fish (hunk 14 above)
Comment (bucaran) [inline functions/import.fish]: Caught.

File: oh-my-fish.fish (hunk 13 above)
Comment (bucaran) [inline oh-my-fish.fish]: Because the path is not $fish_path but $fish_custom, of course.

File: functions/_prepend_tree.fish (hunk 12 above)
Comment (bucaran) [inline functions/_prepend_tree.fish]: Noted.

File: oh-my-fish.fish (hunk 13 above)
Comment (bucaran) [inline oh-my-fish.fish]: I just fixed a bug in `_prepend_tree` where the wildcard was not correctly expanding and not running the for loop at all. fish-shell/issues/683

Also and just for the record, it's good to know to have your wildcards clear:

+ `$fish_custom/**` Traverses both files and directories/subdirectories. Everything. 
+ `$fish_custom/**/` Traverses __only__ directories/subdirectories. (This was another bottleneck :smile:)

File: functions/import.fish (hunk 14 above)
Comment (bucaran) [inline functions/import.fish]: This should simply be `$fish_path` as `import` is not aware of anything else but the fish path. Fixed.

File: oh-my-fish.fish (hunk 13 above)
Comment (bucaran) [inline oh-my-fish.fish]: @bpinto 

### `$fish_path` vs `$fish_custom`
OK. It is handled different _because_ the path we are prepending to `$fish_function_path` is __not__ `$fish_path` but `$fish_custom`.

```fish
for custom_file in $fish_custom/**
  _prepend_path $custom_file -d fish_function_path
  switch $custom_file
    case \*.load
      . $custom_file
  end
end
```
We don't need to call `import` for custom because we are already prepending __everything__ in custom to the path and sourcing all `.load` files by default.

> By not using the same handler (import function) we are also not filtering the custom plugins using the fish_plugins env var.

`$fish_plugins` is just a list with plugin names that exist in `~/$fish_path_plugins`. You can not add custom plugins via `$fish_path`. Those you put in `/custom`. Of course, plugins not under source control inside `/plugins` will always load anyway, but that's not what we are talking about here I presume.

## Wildcards
I just fixed a bug in `_prepend_tree` where the wildcard was not correctly expanding and not running the for loop at all. See here also [fish-shell/issues/683](https://github.com/fish-shell/fish-shell/issues/683)

Also and just for the record, it's good to know to have your wildcards clear:

+ `$fish_custom/**` Traverses both files and directories/subdirectories. Everything. 
+ `$fish_custom/**/` Traverses __only__ directories/subdirectories.

**Comment by bucaran** [comment]
This is it. BTW now `import` prints matched directories by default, but you can skip this redirecting `stdout`. This seems the more unix-like approach I think. Everything should be working now. I will squash into one single nice commit and push again or create another PR.

File: oh-my-fish.fish (hunk 13 above)
Comment (bpinto) [inline oh-my-fish.fish]: > We don't need to call import for custom because we are already prepending everything in custom to the path and sourcing all .load files by default.

This is not the existing behaviour. Custom plugins are just like builtin plugins, they are loaded only if you have them defined on `$fish_plugins`.
There is no reason to not add `$fish_custom` to `import`, we are probably going to do something similar in the future once we add a [vundle like system](https://github.com/bpinto/oh-my-fish/issues/259).

Comment (bucaran) [inline oh-my-fish.fish]: Let's recapitulate, there is `$fish_path` (usually `/.oh-my-fish`) and `$fish_custom` (usually `/.oh-my-fish/custom`), these two are _paths_.

There is also a global _list_ of enabled plugins `$fish_plugins`. Plugins (the plugin name) you add to `$fish_plugins` will be handled by `import` during the initial setup. Plugins as well any other user functions (`.load` files too) defined in the `custom` will be added to the path and sourced respectively.

Comment (bpinto) [inline oh-my-fish.fish]:
```
#Only plugins added to $fish_plugins are loaded.
for plugin in $fish_plugins
  _fish_add_plugin $plugin
  _fish_add_completion $plugin
  _fish_source_plugin_load_file $plugin
end

function _fish_add_plugin
  set -l plugin $argv[1]
  set -l plugin_path "plugins/$plugin"

# It loads the plugin from ~/.oh-my-fish
  _prepend_path $fish_path/$plugin_path -d fish_function_path
# It loads the plugin from ~/.oh-my-fish/custom
  _prepend_path $fish_custom/$plugin_path -d fish_function_path
end
```

```
# All files with .load inside custom folder are automatically loaded.
for config_file in $fish_custom/*.load
  . $config_file
end
```

Comment (bucaran) [inline oh-my-fish.fish]: I see what you mean, but the current implementation does not alter this behavior:

```fish
for custom_file in $fish_custom/**
  _prepend_path $custom_file -d fish_function_path
  switch $custom_file
    case \*.load
      . $custom_file
  end
end

Comment (bpinto) [inline oh-my-fish.fish]: We are loading all custom plugin now, if they were defined on `$fish_plugins` or not.

This loop will go through each file and call `_prepend_path` on it. But this function should only be called for directories. So we will be calling this function many times unnecessarily (for each file).

Only `.load` files under `$fish_custom` root need to be searched and loaded. And the rest behaves just like the `import` function.

Comment (bucaran) [inline oh-my-fish.fish]: OK, give me a few minutes.

Comment (bucaran) [inline oh-my-fish.fish]: Let's continue the conversation over the new PR discussion area. This PR is closed.
